security/aws/rds_unencrypted/AWS_Unencrypted_RDS_Instances.pt

```diff
@@ -0,0 +1,493 @@
+    #Uncomment (remove '#') in the below line to enable Delete action.
+    #escalate $delete_unencrypted_RDS_instances_approval
+    check eq(size(data),0)
+  end
+end
+###############################################################################
+# Escalations
+###############################################################################
+
+escalation "report_unencrypted_RDS_instances" do
+  email $param_email
+end
+
+escalation "delete_unencrypted_RDS_instances_approval" do
+  request_approval  do
+    label "Approve Resource Deletion"
+    description "Approve escalation to run RightScale Cloud Workflow to delete unencrypted RDS-instances"
+    parameter "approval_reason" do
+      type "string"
+      label "Reason for Approval"
+      description "[Note]: RDS Instances with 'DB Instance Status' other than 'Available' and RDS instances with 'delete Protection enabled' cannot be deleted"
+    end
+  end
+  run "delete_unencrypted_RDS_instances", data
+end
+
+###############################################################################
+# Cloud Workflow
+###############################################################################
+
+#https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_CreateDBClusterSnapshot.html
+#https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_DescribeDBClusterSnapshots.html
+#https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_DeleteDBInstance.html
+define delete_unencrypted_RDS_instances($data) return $all_responses do
+  $$debug=true
+  $all_responses = []
+  foreach $item in $data do
+    sub on_error: skip do
+      if ($item['dbCluster_Identifier'] != null) && ($item["delete_Protection"] == 'No')
+        #Creating Cluster snapshot manually Since DB instance snapshot cannot be created directly for Aurora cluster DB instance.
+        $create_cluster_snapshot = http_get(
+          url: "https://rds."+$item["region"]+".amazonaws.com/?Action=CreateDBClusterSnapshot&Version=2014-10-31&DBClusterIdentifier="+$item["dbCluster_Identifier"]+"&DBClusterSnapshotIdentifier="+$item["dbCluster_Identifier"]+"-finalSnapshot",
+          "signature": { type: "aws" }
+        )
+        $status = $create_cluster_snapshot['body']['CreateDBClusterSnapshotResponse']['CreateDBClusterSnapshotResult']['DBClusterSnapshot']['Status']
+
+        $statusCheckCount = 0
+        while ($status != "available") && ($statusCheckCount < 10) do
+          sleep(30)
+          $describe_cluster_snapshot = http_get(
+            url: "https://rds."+$item["region"]+".amazonaws.com/?Action=DescribeDBClusterSnapshots&Version=2014-10-31&DBClusterSnapshotIdentifier="+$item["dbCluster_Identifier"]+"-finalSnapshot",
+            "signature": { type: "aws" }
+          )
+          $status = $describe_cluster_snapshot['body']['DescribeDBClusterSnapshotsResponse']['DescribeDBClusterSnapshotsResult']['DBClusterSnapshots']['DBClusterSnapshot']['Status']
+          $statusCheckCount = $statusCheckCount + 1
+        end
```

Comment:
So what if the status at the end of this 5 minute wait period is still not available? You go on and delete the thing anyways, which isn't correct. You need an if block protecting the final delete to check that status equal "available" before doing the delete. Also I don't know how fast rds does these snapshots but 5 minutes may not be enough, I'm not sure though.

Reply:
True.
I added 'if block' protecting the final delete to check that status equal "available" before doing the delete.
It took less than a minute for me to create cluster Snapshot, So I gave it as max 5min. Because while loop keeps on executing if the status does not change.
Now I added the if block and increased the time to max 15min.

Reply:
Yup 15 min sounds like a more reasonable starting point. I'm guessing snapshot speed might depend on the size, how much has changed, etc so will be slow for an in use db. Thanks for handling all these changes, looks much better now.